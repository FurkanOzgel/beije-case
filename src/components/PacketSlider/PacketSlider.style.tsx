import { StyleSheet } from "react-native";

const styles = StyleSheet.create({
    blackPoint: {
        backgroundColor: "rgb(52, 49, 49)",
        width:3,
        height:3,
        borderRadius: 3,
        marginTop:-1.5
    },
    whitePoint: {
        backgroundColor: "#F5F5F2",
        width:3,
        height:3,
        borderRadius: 5,
        marginTop:-1.5
    },
    title: {
        fontSize: 16,
        fontWeight: "400",
        lineHeight:26,
        height: 42,
        color: "rgba(0, 0, 0, 0.9)",
        letterSpacing: -0.16
    },
    intervalContainer: {
        flexDirection: "row",
        justifyContent: "space-between",
    },
    intervalText: {
        fontSize: 14,
        fontWeight: "400",
        lineHeight:22,
        height: 42,
        color: "rgba(0, 0, 0, 0.9)",
        paddingHorizontal: 5
    },
    sliderValue: {
        borderRadius:2,
        backgroundColor: 'rgb(117,117,117)',
        paddingHorizontal:12,
        paddingVertical: 4,
        marginTop: -45,
        marginLeft:-5,
        position: "absolute"
    },
    sliderValueText: {
        color: "white",
        fontWeight: '500',
        fontSize: 14,
        letterSpacing: -0.48,
        
    },
    thumb: {
        height:18,
        width:18,
        backgroundColor: "rgb(52, 49, 49)",
        borderRadius: 9,
        position: "absolute",
        marginTop: -9,
        marginLeft: -9
    },
    notch: {
        backgroundColor:'rgb(117,117,117)',
        height:15,
        width:15, 
        top: -30, 
        transform: [{ rotate: '45deg' }], 
        zIndex: -1 
    }

});

export default styles;