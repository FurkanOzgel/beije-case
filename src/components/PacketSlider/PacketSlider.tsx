export {}
import React from 'react';
import { Text, View } from 'react-native';
import { Slider } from '@react-native-assets/slider';

import { useDispatch, useSelector } from 'react-redux';
import { Product } from '../../context/cart/models';

import styles from './PacketSlider.style';
import { RootState } from '../../context/store';

interface PacketSliderProps {
    title: string;
    minValue: number;
    maxValue: number;
    step: number;
    categoryID: number;
}

function PacketSlider(props: PacketSliderProps): React.JSX.Element {

    const oldValue = useSelector((state: RootState) =>
        (state.cart.products.filter((product: Product) => 
        product.name == props.title)[0]?.oldQuantity || 0));
    const cart = useSelector((state: RootState) => state.cart);

    const dispatch = useDispatch();

    return (
        <View>
            <Text testID='title' style={styles.title}>{props.title}</Text>
            <View style={{padding: 10}}>
                <Slider
                    testID='slider'
                    value={cart.products.filter((product: Product) => product.name == props.title)[0]?.quantity || 0}
                    minimumValue={0}                  
                    maximumValue={props.maxValue}                 
                    step={props.step}                       
                    minimumTrackTintColor='rgb(52, 49, 49)'      
                    maximumTrackTintColor='rgb(174,171,169)' 
                    thumbStyle={{backgroundColor:'black'}} 
                    trackHeight={4}                   
                    thumbSize={15}                 
                    onValueChange={(value) => {

                        if (value > oldValue) {
                            dispatch({type: 'ADD_PRODUCT', payload: {name: props.title,categoryID: props.categoryID ,quantity: value - oldValue}})
                        } else { 
                            dispatch({type: 'DELETE_PRODUCT', payload: {name: props.title, categoryID: props.categoryID, quantity: oldValue - value}})
                        }
                        
                        dispatch({type: 'UPDATE_OLD_QUANTITY', payload: {name: props.title, oldQuantity: value}})
                        }} 
                    CustomMark={({value}) => {
                        return(
                            <View>
                                {(value > oldValue) ? <View style={styles.blackPoint}/> : <View style={styles.whitePoint}/>}
                            </View>
                        )
                    }}
                    CustomThumb={({value}) => {
                        return(
                            <View>
                                {(value != props.maxValue && value != props.minValue) ? 
                                <View style={{display:'flex', alignItems:'center'}}>
                                    <View style={styles.sliderValue}>
                                        <Text style={styles.sliderValueText}>{value}</Text>
                                    </View>
                                    <View style={styles.notch}></View>
                                    <View style={styles.thumb}/>
                                </View>
                                :
                                <View style={styles.thumb}/>
                                }
                            
                            </View>
                        )
                    }}/>
            </View>
            <View style={styles.intervalContainer}>
                <Text testID='min-value' style={styles.intervalText}>{props.minValue}</Text>
                <Text testID='max-value' style={styles.intervalText}>{props.maxValue}</Text>
            </View>
        </View>
            
    );
}

export default PacketSlider;
