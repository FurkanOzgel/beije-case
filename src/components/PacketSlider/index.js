export {default} from './PacketSlider.tsx';