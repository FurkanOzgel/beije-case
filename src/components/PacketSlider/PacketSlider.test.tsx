import React from 'react'
import {render, screen} from '@testing-library/react-native'
import { Provider } from 'react-redux';
import configureStore from '../../context/store';
import PacketSlider from './PacketSlider'

const store = configureStore();

const comp = 
    <Provider store={store}>
        <PacketSlider title="Test"
            minValue={0}
            maxValue={100}
            step={10}
            categoryID= {1}/>
    </Provider>


test('renders correctly', () => {
    render(comp)
    expect(comp).toMatchSnapshot()
})

test('renders minValue prop', () => {
    render(comp)
    expect(screen.getByTestId("min-value").children[0]).toBe("0")
})

test('renders maxValue prop', () => {
    render(comp)
    expect(screen.getByTestId("max-value").children[0]).toBe("100")
})

test('renders title prop', () => {
    render(comp)
    expect(screen.getByTestId("title").children[0]).toBe("Test")
})
