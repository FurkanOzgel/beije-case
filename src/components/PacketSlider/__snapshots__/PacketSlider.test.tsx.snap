// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Provider
  store={
    {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <PacketSlider
    categoryID={1}
    maxValue={100}
    minValue={0}
    step={10}
    title="Test"
  />
</Provider>
`;
